### ISSUE TITLE
Questions about running the code of stage1.py

### ISSUE THREAD

**Issue by AIforScien** [acharkq/MolCA#2]
Hi,

I hope this message finds you well. I wanted to extend my warmest congratulations on the acceptance of your paper in EMNLP.

I download the file from the repo and set up the environment as yaml file. However, I can't run the code using two gpus with 'python stage1.py --root 'data/PubChem324k' --gtm --lm --devices '0,1' --mode train --filename stage1 --rerank_cand_num 128 --num_query_token 8 --tune_gnn'. Therefore, I change the code to one single gpu --device '1' but I get very low results of retrieval performance by using the trained model from stage1. For example, I get "test_fullset_g2t_rec20     │       79.94999694824219", "test_fullset_t2g_rec20     │       80.05000305175781" on the pcdes dataset when I use the checkpoint from the last epoch I train the model. Can you provide the hyperparameter to obtain the retrieval results as reported in the paper with one single gpu? Thanks for your help.

Best

**Comment by acharkq** [comment]
Hi,

Thanks for your interest in our work. For your question:

```
However, I can't run the code using two gpus with 'python stage1.py --root 'data/PubChem324k' --gtm --lm --devices '0,1' --mode train --filename stage1 --rerank_cand_num 128 --num_query_token 8 --tune_gnn'
```

I have updated the code base. Now you should be able to run evaluation using a single GPU with the following command:

```
python stage1.py --root 'data/kv_data' --gtm --lm --devices '[0]'  --filename pcdes_evaluation --init_checkpoint "all_checkpoints/share/stage1.ckpt" --rerank_cand_num 128 --num_query_token 8 --match_batch_size 64 --mode eval
```

Note that, training stage 1 using a single GPU is still not supported, because it would require a significant revision on the code base. It will also require some hyperparameter tuning because using a single GPU will decrease the batch size, which can hurt the contrastive learning performance.

Also, may I know the reason that you cannot run the code with two GPUs? Is it because of some bugs in our current code base?

**Comment by AIforScien** [comment]
Yes, when I run the code with gpu '0,1'. I get the following error warnings.warn('resource_tracker: There appear to be %d '
Bus error (core dumped). Btw, I change the code in the function contrast_global "change rank = dist.get_rank() to rank=0" to support the experiments.

**Comment by acharkq** [comment]
Thanks. Can you provide more complete error information for your running of two GPUs? Can you copy paste the entire error message?

**Comment by AIforScien** [comment]
Thanks for your help. The end of the output and error are:

```
Using bfloat16 Automatic Mixed Precision (AMP)
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
/root/anaconda3/envs/molca/lib/python3.8/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 4 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
Bus error (core dumped)
```

It didn't show too many error messages.

**Comment by acharkq** [comment]
Hi Bro,

As I searched the internet, the problem is very likely caused by the lack of memory. See this [link](https://stackoverflow.com/questions/64515797/there-appear-to-be-6-leaked-semaphore-objects-to-clean-up-at-shutdown-warnings-w)
